### DISCUSSION TITLE
Can't understand the new behaviour of `useLoaderData`

### DISCUSSION THREAD

**Discussion by nullndr** [remix-run/remix#3927] [Q&A] [answered]
I have some problems with the latest update of `useLoaderData`

Here is my loader:

```ts
export const loader = async () => {
  return {
    requestedDomains: [
      { domain: "example.com", status: "pending" },
      { domain: "saulgoodman.org", status: "approved" },
      { domain: "example.org", status: "rejected", reason: "Not allowed" },
    ],
  };
};
```

And

```ts
const { requestedDomains } = useLoaderData<typeof loader>();
```

`requestedDomains` is of type:

```ts
(SerializeObject<{
    domain: string;
    status: string;
    reason?: undefined;
}> | SerializeObject<{
    domain: string;
    status: string;
    reason: string;
}>)[]
```

Shouldn't it be 

```ts
SerializeObject<{
    domain: string;
    status: string;
    reason?: string;
}>[]
```
?

This behaviour breaks my code because the type should be 

```ts
{
    domain: string;
    status: string;
    reason?: string;
}[]
```

**Comment by kiliman** [comment]
If you want the new type inference to work, you have to do the following:
```js
// Type your loader args
export const loader = async ({ request } : LoaderArgs) => {}

// return your values with `json` helper
return json({ requestedDomains: [{}] })
```

**Reply by nullndr** [reply]
But if I don't use `request`?

Do I have to include it anyway?

**Reply by nullndr** [reply]
However, it still doesn't work.

The type continues to be:

```ts
(SerializeObject<{
    domain: string;
    status: string;
    reason?: undefined;
}> | SerializeObject<{
    domain: string;
    status: string;
    reason: string;
}>)[]
```

**Reply by nullndr** [reply]
I had to pass the `LoaderData` to the `json()` call in order to make it work.

Not sure if this is the correct way, the new update should delete the use of `LoaderData`, right?

**Reply by kiliman** [reply]
Hmm... looks like Remix type inference breaks down with optional properties.

**Reply by nullndr** [reply]
@killman Should I report a bug?

**Reply by kiliman** [reply]
Yes, I don't think it's intentional. I do know they want to use an external lib for that vs home grown, so perhaps that'll be fixed in a future update. https://github.com/remix-run/remix/pull/3907#pullrequestreview-1061765449

**Reply by nullndr** [reply]
Thank you

**Reply by kiliman** [reply]
In the meantime, you can checkout `remix-typedjson`. It will support your use case because it doesn't try to convert your type to the JSON serialized type. Instead, it will convert your JSON data back into the original type (if needed, does nothing if you don't use non-JSON types).

https://github.com/kiliman/remix-typedjson

<img width="677" alt="image" src="https://user-images.githubusercontent.com/47168/183093714-5d4856b6-8395-44a8-ab97-6b25fcd7d8ab.png">

<img width="1018" alt="image" src="https://user-images.githubusercontent.com/47168/183093753-b60ceaba-178e-4afe-9e38-660305e471cd.png">

**Comment by pcattori** [comment, marked as answer]
As described in [this comment](https://github.com/remix-run/remix/issues/3931#issuecomment-1755870135), the behavior you are seeing is inherent to Typescript and is not a Remix issue.
